File: 09-backtracking/00039-Combination_Sum/00039-combination_sum.go
// 39: Combination Sum
// https://leetcode.com/problems/combination-sum/

package main

import "fmt"


func backtrack(candidates []int, target int, start int, current []int, result *[][]int) {
	if start >= len(candidates) || target < 0 {return}

	if target == 0 {
		t := make([]int, len(current))
		copy(t, current)
		*result = append(*result, t)
		return
	}
	
	current = append(current, candidates[start])
	backtrack(candidates, target - candidates[start], start, current, result)
	current = current[:len(current)-1]
	backtrack(candidates, target, start + 1, current, result)
}


// SOLUTION
func combinationSum(candidates []int, target int) [][]int {
    var current []int
	var result [][]int
	backtrack(candidates, target, 0, current, &result)
	return result
}


func main() {
    // INPUT
    candidates := []int{2,3,6,7}
	target := 7

    // OUTPUT
    result := combinationSum(candidates, target)

	fmt.Println(result)
}

File: 09-backtracking/00039-Combination_Sum/00039-combination_sum.py
# 39: Combination Sum
# https://leetcode.com/problems/combination-sum/


from typing import List

class Solution:
    # SOLUTION
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        current = []

        def backtrack(target: int, start: int) -> None:
            if start >= len(candidates) or target < 0 : return

            if target == 0:
                result.append(current.copy())
                return

            current.append(candidates[start])
            backtrack(target - candidates[start], start)
            current.pop()
            backtrack(target, start + 1)

        backtrack(target, 0)
        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    candidates = [2,3,6,7]
    target = 7

    # OUTPUT
    result = o.combinationSum(candidates, target)

    print(result)
